src/main/options.js

```diff
@@ -118,6 +120,53 @@ function inferParser(filepath, plugins) {
   const filepathParts = normalizePath(filepath).split("/");
   const filename = filepathParts[filepathParts.length - 1].toLowerCase();
 
+  let shebang = null;
+  const getShebang = () => {
```

Comment:
I’m not sure what you mean here. Given that this PR is about looking at the shebang, how could it be unrelated to `fs`? Reading (the first line of) the file is exactly what has to be done to check shebangs. Do you mean that the shebang extractions should happen in some module other than `options.js`? Maybe there’s some cleaner place (after all I don’t know the project), but it seems to me that there’s a case for doing it here, because it can be done in *only* cases where the parser *cannot* be inferred from the extension, and I wouldn’t want to do unnecessary FS accesses.

Reply:
> Do you mean that the shebang extractions should happen in some module other than options.js?

Yes, IMO it should be in `resolveConfig` API (`src/config/resolve-config.js`). `options.js` is part of the core module, which is also used in the browser.

What I mean is something like:

```js
// config/resolve-config.js

function resolveConfig(filepath, opts) {
  const interpreter = filepathNoExtname
    ? inferInterpreter(fs.readFirstLine(filepath))
    : null;

  // ...

  return { ...resolvedOptions, interpreter };
}
```

```js
// main/options.js

function inferParser(filepath, interpreter, plugins) {

  // ...

  return inferredParser;
}
```

Reply:
OK, I’ll take a look at that—hopefully tomorrow. This is what happens when non-members try to contribute, I guess… :P I can see why you would not want this part of the browser (I didn’t even know Prettier had any mode of running in the browser!). If I move the check to `resolve-config.js`, will it still apply to both API and CLI invocations?

Reply:
> If I move the check to resolve-config.js, will it still apply to both API and CLI invocations?

Yes, it should. We use the `resolveConfig` API in the CLI.

Reply:
I guess this is what you're looking for:
https://github.com/prettier/prettier/blob/d7e96adac0b089835ab8444846a764c869f847dc/tests_integration/__tests__/file-info.js#L126-L131

Reply:
> > Do you mean that the shebang extractions should happen in some module other than options.js?
> 
> Yes, IMO it should be in `resolveConfig` API (`src/config/resolve-config.js`). `options.js` is part of the core module, which is also used in the browser.

Having found some time to poke at it, I’m not sure why you feel that `resolve-config.js` is the place for this. For one thing, it seems not to be invoked at all by `prettier.getFileInfo()`, and for another, it’s disabled if `prettier` is invoked with `--no-config`. Surely, a user who passes `--no-config` does not expect to thereby disable part of the built-in file type inference! If anything, I guess I could see a case for moving it to `get-file-info.js`, which is where you’d expect to…get file info.

But I’m not convinced, because that still leaves `options.inferParser()` with only *part* of the parser inference logic, which seems like it would cause a maintenance headache. It’s frankly not of great importance to me, since I’m not a maintainer, but I know if I worked regularly on a project, I’d expect `inferParser()` to do its damnedest to infer which parser to use!

Now, my main concern personally is just to get my PR accepted so we can use Prettier for our CLI scripts, so if you say move it to `resolveConfig()` after all, I will and I won’t complain; but right now, `inferParser()` feels like the most logical place to keep it.

-----

I did add a simple test to infer `"babylon"` from `#!/usr/bin/env node`.

Reply:
You're right, I forgot about it. Though I still don't like fs to be used in `options.js`, but it should be fine for now as a temporary solution, we probably need to refactor it later for better structure.

Could you extract the `getInterpreter` function to the root of this file so that it'd be easier to read?

```js
function inferParser(...) {
  // ...

  const interpreter = getInterpreter();

  // ...
}

function getInterpreter(...) {
  // ...
}
```

Reply:
Now *that* can do guilt-free the next time I find myself in a suitable dev enviroment. :P